gpbft/gpbft.go

```diff
@@ -341,120 +343,130 @@ func (i *instance) tryCompletePhase() error {
-		DECIDE_PHASE: {
-			COMMIT_PHASE: {math.MaxUint64, msg.Vote.Value},
-		},
-	}
+		// Check every remaining field of the justification, according to the phase requirements.
+		// This map goes from the message phase to the expected justification phase(s),
+		// to the required vote values.
+		// Anything else is disallowed.
+		expectations := map[Phase]map[Phase]struct {
+			Round uint64
+			Value ECChain
+		}{
+			// CONVERGE is justified by a strong quorum of COMMIT for bottom,
+			// or a strong quorum of PREPARE for the same value, from the previous round.
+			CONVERGE_PHASE: {
+				COMMIT_PHASE:  {msg.Vote.Round - 1, ECChain{}},
+				PREPARE_PHASE: {msg.Vote.Round - 1, msg.Vote.Value},
+			},
+			// COMMIT is justified by strong quorum of PREPARE from the same round with the same value.
+			COMMIT_PHASE: {
+				PREPARE_PHASE: {msg.Vote.Round, msg.Vote.Value},
+			},
+			// DECIDE is justified by strong quorum of COMMIT with the same value.
+			// The DECIDE message doesn't specify a round.
+			DECIDE_PHASE: {
+				COMMIT_PHASE: {math.MaxUint64, msg.Vote.Value},
+			},
+		}
 
-	if expectedPhases, ok := expectations[msg.Vote.Step]; ok {
-		if expected, ok := expectedPhases[msg.Justification.Vote.Step]; ok {
-			if msg.Justification.Vote.Round != expected.Round && expected.Round != math.MaxUint64 {
-				return fmt.Errorf("message %v has justification from wrong round %d", msg, msg.Justification.Vote.Round)
-			}
-			if !msg.Justification.Vote.Value.Eq(expected.Value) {
-				return fmt.Errorf("message %v has justification for a different value: %v", msg, msg.Justification.Vote.Value)
+		if expectedPhases, ok := expectations[msg.Vote.Step]; ok {
+			if expected, ok := expectedPhases[msg.Justification.Vote.Step]; ok {
+				if msg.Justification.Vote.Round != expected.Round && expected.Round != math.MaxUint64 {
+					return fmt.Errorf("message %v has justification from wrong round %d", msg, msg.Justification.Vote.Round)
+				}
+				if !msg.Justification.Vote.Value.Eq(expected.Value) {
+					return fmt.Errorf("message %v has justification for a different value: %v", msg, msg.Justification.Vote.Value)
+				}
+			} else {
+				return fmt.Errorf("message %v has justification with unexpected phase: %v", msg, msg.Justification.Vote.Step)
 			}
 		} else {
-			return fmt.Errorf("message %v has justification with unexpected phase: %v", msg, msg.Justification.Vote.Step)
+			return fmt.Errorf("message %v has unexpected phase for justification", msg)
 		}
-	} else {
-		return fmt.Errorf("message %v has unexpected phase for justification", msg)
-	}
-	return i.verifyJustification(msg.Justification)
-}
 
-func (i *instance) verifyJustification(justification *Justification) error {
-	power := NewStoragePower(0)
-	signers := make([]PubKey, 0)
-	if err := justification.Signers.ForEach(func(bit uint64) error {
-		if int(bit) >= len(i.powerTable.Entries) {
-			return fmt.Errorf("invalid signer index: %d", bit)
+		// Check justification signature.
```

Comment:
What's the rationale for not calling `verifyJustification` anymore?

Reply:
It's still doing the same work, so you mean rationale for not breaking out verifyJustification into a separate function?

It's a close thing, and low impact, so this is all just a matter of design sense in the moment. This change makes it clear that justification is just another part of message validation, rather than different like it used to be with implicit justification. The same question about breaking out a function might be asked of each of checking the vote semantics, the vote signature, the justification semantics, and justification signature. I did consider breaking them all out, but:
- this is just straight line code with very little branching
- the functions are all part of a logical single operation of validation, and don't mean much outside that; they're not re-usable except in another context that also calls all of them
- the functions would each have only one call site
- the functions should be placed spatially next to each other to reflect their one-ness
- breaking out functions would just make the code take more vertical space and add the noise of parameters and return values.

If someone else had done this change and broken them all up, I wouldn't object strongly. But I decided to go with the simple straight line code since I couldn't see any advantage to factoring out.